core/src/core/transaction.rs

```diff
@@ -456,6 +456,33 @@ impl Transaction {
 		Ok(())
 	}
 
+	/// Calculate transaction weight
+	pub fn tx_weight(&self) -> u32 {
+		Transaction::weight(
+			self.inputs.len(),
+			self.outputs.len(),
+			self.input_proofs_count(),
+		)
+	}
+
+	/// Collect input's Merkle proofs
+	pub fn input_proofs_count(&self) -> usize {
+		self.inputs
+			.iter()
+			.filter(|i| i.merkle_proof.is_some())
+			.count()
+	}
+
+	/// Calculate transaction weight from transaction details
+	pub fn weight(input_len: usize, output_len: usize, proof_len: usize) -> u32 {
+		let mut tx_weight =
+			-1 * (input_len as i32) + (4 * output_len as i32) + (proof_len as i32) + 1;
```

Comment:
I think a proof should count for _at least_ the equivalent of an output.
But I'd be good with merging this in now to get the structure of this in place, we can always tweak/adjust weights later.